bundles/org.openhab.ui/web/src/components/widgets/widget-order.js

```diff
@@ -1,11 +1,46 @@
-
 export function compareItems (i1, i2) {
+  const isLocationContext = i1.metadata && i1.metadata.semantics && i1.metadata.semantics.value.indexOf('Location') === 0 &&
+    i2.metadata && i2.metadata.semantics && i2.metadata.semantics.value.indexOf('Location') === 0
+
+  // Compare widgetOrder
   const order1 = (i1.metadata && i1.metadata.widgetOrder) ? i1.metadata.widgetOrder.value : Infinity
   const order2 = (i2.metadata && i2.metadata.widgetOrder) ? i2.metadata.widgetOrder.value : Infinity
-  if (order1 === order2) {
-    const nameOrLabel1 = i1.label || i1.name
-    const nameOrLabel2 = i2.label || i2.name
-    return nameOrLabel1.localeCompare(nameOrLabel2)
+  const widgetOrder = (order1 !== order2) ? ((order1 < order2) ? -1 : 1) : 0
+
+  // When comparing Location Items compare on ancestors first
+  if (!isLocationContext && widgetOrder !== 0) return widgetOrder
+
+  // Compare ancestors in Model (starting at the top level)
+  let modelOrder = 0
+  if (i1.modelPath && i2.modelPath) {
+    const minDepth = Math.min(i1.modelPath.length, i2.modelPath.length) // Compare shortest path
+    for (let d = 0; d < minDepth; d++) {
+      if (i1.modelPath[d] === i2.modelPath[d]) continue
+
+      // widgetOrder comparison...
+      const order1 = (i1.modelPath[d].metadata && i1.modelPath[d].metadata.widgetOrder) ? i1.modelPath[d].metadata.widgetOrder.value : Infinity
+      const order2 = (i2.modelPath[d].metadata && i2.modelPath[d].metadata.widgetOrder) ? i2.modelPath[d].metadata.widgetOrder.value : Infinity
+      if (order1 !== order2) {
+        modelOrder = (order1 < order2) ? -1 : 1
+        break
+      }
+
+      // ... or lexicographical comparison
+      const nameOrLabel1 = i1.modelPath[d].label || i1.modelPath[d].name
+      const nameOrLabel2 = i2.modelPath[d].label || i2.modelPath[d].name
+      modelOrder = nameOrLabel1.localeCompare(nameOrLabel2)
+      if (modelOrder !== 0) break
+    }
+
+    // A parent compares greater than its children
+    if (modelOrder === 0 && (i1.modelPath.length !== i2.modelPath.length)) modelOrder = (i1.modelPath.length < i2.modelPath.length) ? -1 : 1
+
+    if (modelOrder !== 0) return modelOrder
```

Comment:
Could you please extract this to its own function to make the "default" case easier to parse?

Reply:
Just did it. It might also have some uses elsewhere as a standalone function.